File: CSC310-Cpp/_solutions/Module11-Graph-Algorithms/debug01/bfs_distances.cpp
// bfs_distances.cpp
// Breadth-first search (CLRS 20.2) on an undirected graph. Computes the
// distance (fewest edges) from a source to every vertex, the BFS-tree parent
// of each vertex, and counts how many times a vertex was placed in the queue.
#include <iostream>
#include <queue>
#include <string>
#include <vector>

const std::vector<std::string> names = {"r", "s", "t", "u", "v", "w", "x", "y"};

struct BfsResult {
    std::vector<int> dist;
    std::vector<int> parent;
    long enqueues;
};

BfsResult bfs(const std::vector<std::vector<int>>& adj, int source) {
    const int n = static_cast<int>(adj.size());
    BfsResult r{std::vector<int>(n, -1), std::vector<int>(n, -1), 0};
    std::vector<bool> visited(n, false);
    std::queue<int> q;

    r.dist[source] = 0;
    visited[source] = true;
    q.push(source);
    ++r.enqueues;

    while (!q.empty()) {
        int u = q.front();
        q.pop();
        for (int v : adj[u]) {
            if (!visited[v]) {
                visited[v] = true;
                r.dist[v] = r.dist[u] + 1;
                r.parent[v] = u;
                q.push(v);
                ++r.enqueues;
            }
        }
    }
    return r;
}

int main() {
    const int n = 8;
    std::vector<std::vector<int>> adj(n);
    auto edge = [&](const std::string& a, const std::string& b) {
        int u = 0, v = 0;
        for (int i = 0; i < n; ++i) {
            if (names[i] == a) u = i;
            if (names[i] == b) v = i;
        }
        adj[u].push_back(v);
        adj[v].push_back(u);
    };
    edge("r", "s"); edge("r", "v"); edge("s", "w"); edge("w", "t"); edge("w", "x");
    edge("t", "x"); edge("t", "u"); edge("x", "u"); edge("x", "y"); edge("u", "y");

    BfsResult r = bfs(adj, 1);   // source s
    std::cout << "BFS from s\n";
    for (int v = 0; v < n; ++v) {
        std::cout << "  " << names[v] << ": dist " << r.dist[v] << ", parent "
                  << (r.parent[v] < 0 ? "-" : names[r.parent[v]]) << '\n';
    }
    std::cout << "vertices enqueued: " << r.enqueues << '\n';
    return 0;
}

File: CSC310-Python/Module11-Graph-Algorithms/debug01/bfs_distances.py
"""Breadth-first search (CLRS Section 20.2).

bfs(adj, s) computes the shortest-path distance (number of edges) from
the source s to every reachable vertex, together with the BFS tree, and
counts how many times a vertex was placed on the queue.  The main block
runs the search on the undirected example graph of CLRS Section 20.2
from source s.
"""

from collections import deque


def bfs(adj, source):
    dist = {v: None for v in adj}
    parent = {v: None for v in adj}
    visited = {v: False for v in adj}
    enqueued = 0

    dist[source] = 0
    queue = deque([source])
    enqueued += 1
    order = []
    while queue:
        u = queue.popleft()
        if visited[u]:
            continue
        visited[u] = True
        order.append(u)
        for v in adj[u]:
            if not visited[v]:
                dist[v] = dist[u] + 1
                parent[v] = u
                queue.append(v)
                enqueued += 1
    return dist, parent, order, enqueued


def path_to(parent, v):
    path = []
    while v is not None:
        path.append(v)
        v = parent[v]
    return "->".join(reversed(path))


def main():
    # The BFS example graph of CLRS Section 20.2 (undirected)
    edges = [("r", "v"), ("r", "s"), ("s", "w"), ("w", "t"), ("w", "x"),
             ("t", "x"), ("t", "u"), ("x", "u"), ("x", "y"), ("u", "y")]
    adj = {v: [] for v in "rstuvwxy"}
    for a, b in edges:
        adj[a].append(b)
        adj[b].append(a)

    dist, parent, order, enqueued = bfs(adj, "s")
    print("dequeue order :", " ".join(order))
    print("times enqueued:", enqueued, "(vertices:", len(adj), ")")
    print()
    for v in sorted(adj):
        print(f"  {v}: dist {dist[v]}   path {path_to(parent, v)}")


if __name__ == "__main__":
    main()
